<a href="https://colab.research.google.com/github/UadrianSP/am-turma5/blob/main/TratamentoCensus_aula0901.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pipeline de Analise de Dados [Census]

## Importação de Bibliotecas

*   Pandas
*   Numpy
*   MatPlotLib
*   SeaBorn



In [14]:
!pip install pandas
!pip install numpy
!pip install matplotlib
!pip install seaborn

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [16]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [17]:
import os

def localizar_arquivo(nome_arquivo, pasta_inicial='/content/drive'):
    print(f"Buscando por '{nome_arquivo}' em {pasta_inicial}...")
    for raiz, dirs, arquivos in os.walk(pasta_inicial):
        if nome_arquivo in arquivos:
            caminho_completo = os.path.join(raiz, nome_arquivo)
            print(f"\n✅ ARQUIVO ENCONTRADO! Caminho correto:")
            print(caminho_completo)
            return caminho_completo
    print(f"\n❌ Arquivo '{nome_arquivo}' não encontrado no Drive.")
    return None

caminho_real = localizar_arquivo('census.csv')

Buscando por 'census.csv' em /content/drive...

✅ ARQUIVO ENCONTRADO! Caminho correto:
/content/drive/MyDrive/am-t6/census.csv


In [18]:
from google.colab import files
import os

print("Por favor, selecione o arquivo 'census.csv' do seu computador:")
uploaded = files.upload()

if 'census.csv' in uploaded:
    print("\n✅ Arquivo carregado com sucesso!")
    # Caminho local no Colab
    caminho_local = '/content/census.csv'
    # Dica: use este caminho no pd.read_csv(caminho_local)
else:
    print("\n❌ O arquivo enviado não tem o nome 'census.csv'. Por favor, tente novamente.")

Por favor, selecione o arquivo 'census.csv' do seu computador:


Saving census.csv to census (1).csv

❌ O arquivo enviado não tem o nome 'census.csv'. Por favor, tente novamente.


In [19]:
import os

# Caminho que você está tentando acessar
caminho_alvo = '/content/drive/MyDrive/am-t6/'

if os.path.exists(caminho_alvo):
    print(f"Arquivos encontrados em {caminho_alvo}:")
    print(os.listdir(caminho_alvo))
else:
    print(f"A pasta {caminho_alvo} não foi encontrada. Listando o conteúdo de 'MyDrive' para ajudar:")
    print(os.listdir('/content/drive/MyDrive/'))

Arquivos encontrados em /content/drive/MyDrive/am-t6/:
['census.csv', 'credit_data.csv']


In [20]:
import os

# Vamos verificar a raiz do drive montado
raiz_drive = '/content/drive'

if os.path.exists(raiz_drive):
    print(f"Pastas encontradas em {raiz_drive}:")
    try:
        print(os.listdir(raiz_drive))
        # Tenta encontrar o arquivo recursivamente se a lista acima não ajudar
        print("\nBuscando 'census.csv' em subpastas...")
        for root, dirs, files in os.walk(raiz_drive):
            if 'census.csv' in files:
                print(f"ARQUIVO ENCONTRADO EM: {os.path.join(root, 'census.csv')}")
    except Exception as e:
        print(f"Erro ao acessar subpastas: {e}")
else:
    print("O drive não parece estar montado em /content/drive. Tente rodar a célula de mount novamente.")

Pastas encontradas em /content/drive:
['.shortcut-targets-by-id', 'MyDrive', '.Trash-0', '.Encrypted']

Buscando 'census.csv' em subpastas...
ARQUIVO ENCONTRADO EM: /content/drive/MyDrive/am-t6/census.csv


## Carregamento dos Dados da Base

In [21]:
base_census = pd.read_csv('/content/drive/MyDrive/am-t6/census.csv')

In [22]:
base_census

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32557,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32558,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K
32559,22,Private,201490,HS-grad,9,Never-married,Adm-clerical,Own-child,White,Male,0,0,20,United-States,<=50K


In [23]:
base_census.describe()

,age,final-weight,education-num,capital-gain,capital-loos,hour-per-week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


## Caracteristicas da Base
- População amostral: 32.561
- Minorias importantes: 28 e 90, vol. Ed.=9
e 16,
ht= 40 e 99
- Media amostral: idade = 38.58, Ed = 10,
Gain. 1077.65, Loss = 87.30, ht= 40.44
- Desvio Padrão: [limites]: idade = 24.94, Ed. = 7.
51, Gain. = -6757.64 Loss= -315.66, ht= 28,9

In [24]:
import plotly.express as px
import pandas as pd

# Usando o caminho do arquivo carregado localmente no Colab
base_census = pd.read_csv('/content/census.csv')
grafico_classe_wa = px.treemap(base_census, path=['workclass', 'age'])
grafico_classe_wa.show()

In [25]:
grafico_fineture_eri = px.parallel_categories(base_census, dimensions= ['education','race','income'])
grafico_fineture_eri.show()

# Analise de Anomalias de ddos [Censo]

# Analilise de Anomalias de Dados [Census]


**Enunciado: Auditoria e Higienização de Dados Censitários**

Vamos analizar e estruturar a ingestão de um novo pipeline de dados sociodemográficos focado no desenvolvimento de modelos de inteligência artificial. O arquivo **census.csv** foi extraído de um sistema legado, mas o processo automatizado falhou antes mesmo da etapa de modelagem estatística. Como futuros cientistas da computação e engenheiros de dados, a missão de vocês é realizar uma auditoria rigorosa neste arquivo para garantir a integridade da informação.

Dicas de Investigação e Solução

* **Identificação de Linhas Infiltradas**: Ao forçarem o carregamento inicial (pulando as linhas defeituosas do topo), observem o comportamento dos tipos de dados informados pelo `.dtypes`. Se colunas que deveriam ser numéricas (`age` ou `hour-per-week`) foram lidas como `object`, existem *strings* perdidas no meio da massa de dados. O cabeçalho e fragmentos de texto corrompido estão repetidos ao longo do arquivo.
* **Inconsistência Categórica e Caracteres de Escape:** Analisem a distribuição de classes na coluna `income` através do `.value_counts()`. A mesma classe de renda foi acidentalmente dividida em múltiplas categorias distintas devido a falhas de exportação com caracteres de escape textuais (como `\<=50K` versus `<=50K`).
* **Dados Faltantes Não Convencionais:** Avaliem como a ausência de informação foi mascarada pelo sistema de origem. Em bases censitárias clássicas, campos em branco (como `workclass` e `native-country`) frequentemente não aparecem como `NaN` estruturados, mas são substituídos por marcadores categóricos arbitrários, como o caractere `?`.
* **Tratamento de Espaçamento:** A separação por vírgulas pode conter espaços em branco residuais no início das palavras, prejudicando filtros e agrupamentos futuros. O parâmetro `skipinitialspace=True` na leitura do Pandas é um atalho eficiente para padronizar as bordas do texto.

### 1. Auditoria Inicial e Identificação de Problemas
Vamos recarregar os dados usando `skipinitialspace=True` para evitar problemas de filtragem e verificar os tipos de dados (`.dtypes`) para encontrar valores não numéricos em colunas de números.

In [30]:
import pandas as pd

# Carregando com skipinitialspace para remover espaços extras
base_census = pd.read_csv('/content/census.csv', skipinitialspace=True)

print("--- Tipos de Dados ---")
print(base_census.dtypes)

print("\n--- Distribuição da coluna 'income' (Investigando caracteres de escape) ---")
print(base_census['income'].value_counts())

--- Tipos de Dados ---
age                int64
workclass         object
final-weight       int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loos       int64
hour-per-week      int64
native-country    object
income            object
dtype: object

--- Distribuição da coluna 'income' (Investigando caracteres de escape) ---
income
<=50K    24720
>50K      7841
Name: count, dtype: int64


### 2. Limpeza de Dados Infiltrados e Padronização
Se houver strings como 'age' repetidas no meio do arquivo, as colunas numéricas estarão como `object`. Vamos converter para numérico e investigar valores faltantes mascarados (`?`).

In [31]:
# 1. Remover linhas que repetem o cabeçalho (se existirem)
base_census = base_census[base_census['age'] != 'age']

# 2. Converter colunas numéricas que foram lidas como object por erro de exportação
colunas_numericas = ['age', 'final-weight', 'education-num', 'capital-gain', 'capital-loos', 'hour-per-week']
for col in colunas_numericas:
    base_census[col] = pd.to_numeric(base_census[col], errors='coerce')

# 3. Corrigir caracteres de escape na coluna income (ex: \<=50K para <=50K)
base_census['income'] = base_census['income'].str.replace(r'\\', '', regex=True)

# 4. Identificar e tratar o caractere '?' como NaN
import numpy as np
base_census = base_census.replace('?', np.nan)

print("\n--- Valores nulos identificados (após tratar '?') ---")
print(base_census.isnull().sum())

print("\n--- Income após correção ---")
print(base_census['income'].value_counts())

display(base_census.head())


--- Valores nulos identificados (após tratar '?') ---
age                  0
workclass         1836
final-weight         0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loos         0
hour-per-week        0
native-country     583
income               0
dtype: int64

--- Income após correção ---
income
<=50K    24720
>50K      7841
Name: count, dtype: int64


,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


### 3. Higienização Final
Nesta etapa, vamos remover as linhas com valores nulos identificados e eliminar registros duplicados para garantir que o modelo de IA receba apenas dados limpos.

In [32]:
# Removendo valores nulos (NaN) resultantes do tratamento do '?'
base_census_limpa = base_census.dropna()

# Removendo duplicatas
base_census_limpa = base_census_limpa.drop_duplicates()

print(f'Tamanho da base original: {len(base_census)}')
print(f'Tamanho da base após higienização: {len(base_census_limpa)}')
print(f'Total de linhas removidas: {len(base_census) - len(base_census_limpa)}')

# Verificação final da coluna income
print('\n--- Distribuição Final de Renda ---')
print(base_census_limpa['income'].value_counts())

display(base_census_limpa.info())

Tamanho da base original: 32561
Tamanho da base após higienização: 30139
Total de linhas removidas: 2422

--- Distribuição Final de Renda ---
income
<=50K    22633
>50K      7506
Name: count, dtype: int64
<class 'pandas.core.frame.DataFrame'>
Index: 30139 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30139 non-null  int64 
 1   workclass       30139 non-null  object
 2   final-weight    30139 non-null  int64 
 3   education       30139 non-null  object
 4   education-num   30139 non-null  int64 
 5   marital-status  30139 non-null  object
 6   occupation      30139 non-null  object
 7   relationship    30139 non-null  object
 8   race            30139 non-null  object
 9   sex             30139 non-null  object
 10  capital-gain    30139 non-null  int64 
 11  capital-loos    30139 non-null  int64 
 12  hour-per-week   30139 non-null  int64 
 13  native-country  30139 non-

None

### 3. Higienização Final
Nesta etapa, vamos remover as linhas com valores nulos identificados e eliminar registros duplicados para garantir que o modelo de IA receba apenas dados limpos.

In [33]:
# Removendo valores nulos (NaN) resultantes do tratamento do '?'
base_census_limpa = base_census.dropna()

# Removendo duplicatas
base_census_limpa = base_census_limpa.drop_duplicates()

print(f"Tamanho da base original: {len(base_census)}")
print(f"Tamanho da base após higienização: {len(base_census_limpa)}")
print(f"Total de linhas removidas: {len(base_census) - len(base_census_limpa)}")

# Verificação final da coluna income
print("\n--- Distribuição Final de Renda ---")
print(base_census_limpa['income'].value_counts())

display(base_census_limpa.info())

Tamanho da base original: 32561
Tamanho da base após higienização: 30139
Total de linhas removidas: 2422

--- Distribuição Final de Renda ---
income
<=50K    22633
>50K      7506
Name: count, dtype: int64
<class 'pandas.core.frame.DataFrame'>
Index: 30139 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30139 non-null  int64 
 1   workclass       30139 non-null  object
 2   final-weight    30139 non-null  int64 
 3   education       30139 non-null  object
 4   education-num   30139 non-null  int64 
 5   marital-status  30139 non-null  object
 6   occupation      30139 non-null  object
 7   relationship    30139 non-null  object
 8   race            30139 non-null  object
 9   sex             30139 non-null  object
 10  capital-gain    30139 non-null  int64 
 11  capital-loos    30139 non-null  int64 
 12  hour-per-week   30139 non-null  int64 
 13  native-country  30139 non-

None